File: DynamicProgramming/wordBreak1.py
""" https://leetcode.com/problems/word-break/ """

#using recursion exponetial in time
class Solution(object):
    #time complexity : O(W^N), where W is the length of wordDict and N is the length of string
    def helperRec(self, wordDict, i, s):
        #base case
        if i == len(s):
            return True
        
        #rec case
        for word in wordDict:
            l = len(word)
            if word == s[i: i + l] and self.helperRec(wordDict, i + l, s):
                return True
            
        return False
        
    def wordBreak(self, s, wordDict):
    
        return self.helperRec(wordDict, 0, s)


#using top down dp aka memoization
class Solution(object):

    def helperRec(self, wordDict, i, s, dp):
        #base case
        if i == len(s):
            dp[i] = True
        
        if i in dp:
            return dp[i]

        #rec case
        for word in wordDict:
            l = len(word)
            if word == s[i: i + l] and self.helperRec(wordDict, i + l, s, dp):
                dp[i + l] = True
                return True
            
        dp[i] = False
        return False
        
    def wordBreak(self, s, wordDict):
        
        dp = {}
        return self.helperRec(wordDict, 0, s, dp)
        

File: DynamicProgramming/wordBreak1.cpp
#include <iostream>
#include<vector>
#include <unordered_map> //use this
using namespace std;

class Solution {
public:
    bool solveMem(string s, vector<string>& wordDict, unordered_map <int,bool> &dp, int i){
        
        //base case
        if(i == s.length()) return true;
        
        
        //memoization 
        if (dp.find(i) != dp.end()) return dp[i];
        
        //rec case
        for(auto word : wordDict){
            
            if(word == s.substr(i, word.length()) and solveMem(s, wordDict, dp, i+word.length())){
                
                dp[i + word.length()] = true;
                return true;
            }
        }
        
        dp[i] = false;
        return false;
        
    }
    
    bool wordBreak(string s, vector<string>& wordDict) {
        
        //unordered_map <int, bool> dp;
        //return solveMem(s, wordDict, dp, 0);
        vector<bool>dp(s.length() + 1, false);
        dp[s.length()] = true;
        
        for(int i = s.length()-1; i >= 0; i--){
            
            for(auto word : wordDict){
                
                if(word.length() + i <= s.length() and (word == s.substr(i, word.length()))){
                    
                    dp[i] = dp[i + word.length()];
                
                }
                
                if (dp[i]) break;
            }
            
        }  
        
        return dp[0];      
    }
};